starcheck/src/lib/Ska/Parse_CM_File.pm

```diff
@@ -61,93 +61,120 @@ sub TLR_load_segments{
 
-    # if the most recent/last entry in the dither file has a timestamp newer than
-    # the first entry in the load
-    my $dither_time_violation = ($dh_time[-1] >= $bs_arr->[0]->{time});
-    if ($dither_time_violation){
-      return ($dither_time_violation, undef);
+    # If the most recent/last entry in the dither file has a timestamp newer than
+    # the first entry in the load, return string error and undef dither history.
+    if ($dh_time[-1] >= $bs_arr->[0]->{time}){
+       return ("Dither history runs into load\n", undef);
     }
 
+    # Confirm that last state matches kadi continuity ENAB/DISA.
+    # Otherwise, return string error and undef dither history.
+    if ($kadi_dither->{'dither'} ne $dh_state[-1]){
+        return ("Dither status in kadi commands does not match DITHER history\n", undef);
+    }
+
+
+
     # Now make an array of hashes as the final output.  Keep track of where the info
-    # came from to assist in debugging.
+    # came from to assist in debugging ('source' key)
     my @dither;
     my $bs_start = $bs_arr->[0]->{time};
     my $r2a = 3600. * 180. / 3.14159265;
     my $pi = 3.14159265;
 
-    my $dither_state;
-    my $dither_ampl_p;
-    my $dither_ampl_y;
-    my $dither_period_p;
-    my $dither_period_y;
- 
-    foreach (0 .. $#state) {
-      $dither_state = $state[$_] if defined $state[$_];
-      $dither_ampl_p = $params[$_]->{COEFP} * $r2a if defined $params[$_]->{COEFP};
-      $dither_ampl_y = $params[$_]->{COEFY} * $r2a if defined $params[$_]->{COEFY};
-      $dither_period_p = 1 / ($params[$_]->{RATEP} / (2 * $pi))
-         if defined $params[$_]->{RATEP};
-      $dither_period_y = 1 / ($params[$_]->{RATEY} / (2 * $pi))
-         if defined $params[$_]->{RATEP};
-      push @dither, { time => $time[$_],
-		      state => $dither_state,
-		      source => $time[$_] < $bs_start ? 'history' : 'backstop',
-		      ampl_p => $dither_ampl_p,
-		      ampl_y => $dither_ampl_y,
-                      period_p => $dither_period_p,
-                      period_y => $dither_period_y,
-                      tlmsid => $params[$_]->{TLMSID},
-                  };
-  }
-    return ($dither_time_violation, \@dither);
+
+    my $dither_enab = $kadi_dither->{'dither'};
+    my $dither_ampl_p = $kadi_dither->{'dither_ampl_pitch'};
+    my $dither_ampl_y = $kadi_dither->{'dither_ampl_yaw'};
+    my $dither_period_p = $kadi_dither->{'dither_period_pitch'};
+    my $dither_period_y = $kadi_dither->{'dither_period_yaw'};
+
+    push @dither, { 'time' => $kadi_dither->{'time'},
+                    'state' => $dither_enab,
+                    'source' => 'kadi',
+                    'ampl_p' => $dither_ampl_p,
+                    'ampl_y' => $dither_ampl_y,
+                    'period_p' => $dither_period_p,
+                    'period_y' => $dither_period_y};
+
+
+
+    # Build dither states using just kadi initial state and backstop cmds
+    foreach $bs (@{$bs_arr}) {
```

Comment:
There was a loop over the bs_arr that was just making three new arrays and then looping over '@states' from that.  I didn't really see the point in looping twice (except to save one level of indentation in the larger block).